edward/criticisms/evaluate.py

```diff
@@ -291,6 +307,39 @@ def sparse_categorical_crossentropy(y_true, y_pred):
       logits=y_pred, labels=y_true))
 
 
+def multinomial_accuracy(y_true, y_pred):
+  """Multinomial prediction accuracy. `y_true` is a tensor
+  of integers, where the outermost dimension gives a draw
+  from a Multinomial distribution.
+
+  NB: In evaluating the accuracy between two Multinomials
+  results may vary across evaluations. This is because Edward's
+  algorithm for computing `y_pred`, i.e. the Multinomial
+  mode, yields variable results if `any(isinstance(p, float)
+  for p in total_count * probs)` (where `probs` is a vector
+  of the predicted Multinomial probabilities).
+  """
+  y_true = tf.cast(y_true, tf.float32)
+  y_pred = tf.cast(y_pred, tf.float32)
+  return tf.reduce_mean(tf.cast(tf.equal(y_true, y_pred), tf.float32))
+
+
+def kl_divergence(y_true, y_pred):
+  """Kullback-Leibler divergence between two probability distributions. A
+  vector of probabilities for `y_true`.
+
+  Args:
+    y_true: tf.Tensor.
+      Tensor of real values (probabilities) where the values in each row
+      of the outermost dimension sum to 1.
+    y_pred: tf.Tensor.
+      Same as `y_true`, and with the same shape.
+  """
+  y_true = tf.cast(y_true + 1.0, tf.float32)
```

Comment:
`y_true` can plausibly include 0'; this would be problematic in taking the log.

Should I use `+ 1e-6` instead? (I feel like I'm missing something.)

Reply:
The KL divergence for two discrete distributions p to q is a weighted sum of log ratios. Whenever p has zero probability in a term, the corresponding summand is zero. Whenever q has zero probability in a term, p must as well or else the KL divergence is undefined.

Assuming this is `KL(true || predict)`, you might implement this via:
```python
def ...:
  y_true = tf.cast(y_true, tf.float32)
  y_pred = tf.cast(y_pred, tf.float32)
  summand = tf.where(tf.equal(y_true, 0.0), 0.0, y_true * tf.log(y_true / y_pred))
  return tf.reduce_sum(summand)
```

Reply:
Thanks, Dustin.

Is "whenever q has zero probability in a term, p must as well or else the KL divergence is undefined" codified in the formula above? Need it be?

Reply:
I predict the function will return a `tf.Tensor` evaluating to `np.nan` so it might be "implemented"; you can also check this in tests.